### ISSUE TITLE
Folding for CMake incorrectly folds functions and elseif

### ISSUE THREAD

**Issue by ssproessig** [notepad-plus-plus/notepad-plus-plus#7901]
### Description of the Issue`
Example CMake file `example.cmake`

```cmake
function(function1)
endfunction()

function(function2 PARAM)
    if(${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
        list(APPEND options ${COMPILE_OPTIONS_GCC})
    elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL Clang)
        list(APPEND options ${COMPILE_OPTIONS_CLANG})
    elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL MSVC)
        list(APPEND options ${COMPILE_OPTIONS_MSVC})
    endif()

    target_compile_options(${TARGET} PRIVATE ${options})
endfunction()

function(function3 PARAM P2)
endfunction()
```

### Steps to Reproduce the Issue
1.  Copy `example.cmake` into a file
2. Open in Notepad++

### Expected Behavior
- three functions to fold (`function1`, `function2`, `function3`)
- each `if`/`elseif` can be folded

### Actual Behavior
- `elseif` folding does not handle multiple `elseif` correctly (fold from first `elseif` to `endif`)
- `function` folding breaks in `function2` - folding everything until last `endfunction`

### Screenshots
![cmake_folding_1](https://user-images.githubusercontent.com/1576323/73391277-e8d69400-42d7-11ea-9560-84deaf913bdb.png)

![cmake_folding_3](https://user-images.githubusercontent.com/1576323/73391280-e96f2a80-42d7-11ea-8a3d-0141570b9d17.png)

![cmake_folding_2](https://user-images.githubusercontent.com/1576323/73391279-e96f2a80-42d7-11ea-82fa-e2112d28e2cc.png)

### Debug Information
```
Notepad++ v7.8.2   (64-bit)
Build time : Dec  4 2019 - 01:39:49
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1909
OS Build : 18363.592
Plugins : mimeTools.dll NppConverter.dll NppExport.dll 
````

**Comment by KeyArts** [comment]
Hello, 

this error has been bothering me for a while, so I started to investigate it.

I think the solution is quite simple.

The only thing necessary to avoid this error is that the fold level is not incremented on an "elseif"-token.

See my amendment on this:
```
diff --git a/scintilla/lexers/LexCmake.cxx b/scintilla/lexers/LexCmake.cxx
index 06308fbb..06eca33f 100644
--- a/scintilla/lexers/LexCmake.cxx
+++ b/scintilla/lexers/LexCmake.cxx
@@ -85,7 +85,7 @@ static int calculateFoldCmake(Sci_PositionU start, Sci_PositionU end, int foldle
 
     if ( CompareCaseInsensitive(s, "IF") == 0 || CompareCaseInsensitive(s, "WHILE") == 0
          || CompareCaseInsensitive(s, "MACRO") == 0 || CompareCaseInsensitive(s, "FOREACH") == 0
-         || CompareCaseInsensitive(s, "FUNCTION") == 0 || CompareCaseInsensitive(s, "ELSEIF") == 0)
+         || CompareCaseInsensitive(s, "FUNCTION") == 0)
         newFoldlevel++;
     else if ( CompareCaseInsensitive(s, "ENDIF") == 0 || CompareCaseInsensitive(s, "ENDWHILE") == 0
               || CompareCaseInsensitive(s, "ENDMACRO") == 0 || CompareCaseInsensitive(s, "ENDFOREACH") == 0
```

With regards

**Comment by vinsworldcom** [comment]
@KeyArts because this bug is Scintilla related, you'll probably have better luck submitting it to [Scintilla](https://sourceforge.net/p/scintilla/bugs/).  Historically, bugs in Scintilla components are not fixed in Notepad++ until they are fixed in Scintilla and then Notepad++ upgrades to that (or later) version of Scintilla.

Cheers.

**Comment by KeyArts** [comment]
Hi vinsworldcom, 

thank you for suggestion, so i posted my solution also on [sourceforge](https://sourceforge.net/p/scintilla/bugs/2213/#25b2)

**Comment by mpheath** [comment]
@vinsworldcom

This is [Lexilla](https://github.com/ScintillaOrg/lexilla) related. Scintilla is the edit component that you type into, Lexilla is the lexing component that does the language styling and folding. The separation of Scintilla and Lexilla happened in the version 5 series.

@KeyArts

Can you open an issue at [Lexilla](https://github.com/ScintillaOrg/lexilla) with the [file](https://sourceforge.net/p/scintilla/bugs/_discuss/thread/24e6a4bdaf/25b2/attachment/LexCmake.cxx) posted on SourceForge attached. Neil Hodgson will accept files, patches and pull requests, though a file will be easiest for this task. Might be a good idea to include the link to the SourceForge ticket so it can be closed once commited at Lexilla. I could open the issue on your behalf, though I prefer someone with cmake knowledge in case it is needed and it will give you the credit deserved.

I have reviewed the file and the folding test looks good in SciTE. I can post some test files at Lexilla for the [TestLexers](https://github.com/ScintillaOrg/lexilla/tree/master/test) requirements to help the fix to be accepted. Sooner is better so I can co-ordinate with you to get this well setup for Neil to review. Neil announced a break, so may take up to 4 weeks until his involvement.

BTW, your patch here looks backwards, I get:

```diff
--- a/lexilla/lexers/LexCmake.cxx
+++ b/lexilla/lexers/LexCmake.cxx
@@ -88,7 +88,7 @@
 
     if ( CompareCaseInsensitive(s, "IF") == 0 || CompareCaseInsensitive(s, "WHILE") == 0
          || CompareCaseInsensitive(s, "MACRO") == 0 || CompareCaseInsensitive(s, "FOREACH") == 0
-         || CompareCaseInsensitive(s, "FUNCTION") == 0)
+         || CompareCaseInsensitive(s, "FUNCTION") == 0 || CompareCaseInsensitive(s, "ELSEIF") == 0)
         newFoldlevel++;
     else if ( CompareCaseInsensitive(s, "ENDIF") == 0 || CompareCaseInsensitive(s, "ENDWHILE") == 0
               || CompareCaseInsensitive(s, "ENDMACRO") == 0 || CompareCaseInsensitive(s, "ENDFOREACH") == 0

```

to add `+` the ELSEIF condition, not to remove `-` it.

**Comment by KeyArts** [comment]
@mpheath 
Hi Michael, 
I posted the Fix again in ScintillaOrg-Project. https://github.com/ScintillaOrg/lexilla/issues/77
It would be great, if you could add some Testcases. Thank you very much.
With regards
Knut

**Comment by rdipardo** [comment]
This can be closed now that the issue is resolved upstream: ScintillaOrg/lexilla@0422d84
